### PR TITLE
fix search word splitter in JavaScript

### LINKED ISSUE #3150: BugFix to support CJK search
https://github.com/sphinx-doc/sphinx/commit/4ff5cbfaf920700382f809abe5def23e07bd0553

This change will break the CJK search. For Chinese:

```JavaScript
var query = "Hello from 中国 上海";
var tmp = query.split(/\W+/);
// tmp: [ 'Hello', 'from', '' ];
```
Former one:
```JavaScript
var query = "Hello from 中国 上海";
var tmp = query.split(/\s+/);
// tmp: [ 'Hello', 'from', '中国', '上海' ]
```

I suggest to revert the commit.

### PR DESCRIPTION
**Pull request by shibukawa**
Now I made code generator and it works as same as Python.
It fix #3150

### CODE REVIEW

File: utils/compat_regexp_generator.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+console.log('   ... ok\\n')
+''' % js_src
+
+python_src = '''
+"""# -*- coding: utf-8 -*-
+    sphinx.search.jssplitter
+    ~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Provides Python compatible word splitter to JavaScript
+
```
Comment (tk0miya) [inline utils/compat_regexp_generator.py]: Could you add a comment like "DO NOT EDIT. This is generated by compat_regexp_generator.py."?

File: utils/compat_regexp_generator.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
+# -*- coding: utf-8 -*-
```
Comment (tk0miya) [inline utils/compat_regexp_generator.py]: Should this script be installed? What about move this into `utils` directory of root directory?

Comment (tk0miya) [reply utils/compat_regexp_generator.py]: `compat_regexp` is very vague for me.
What about `jssplitter_generator`?

**Review by tk0miya — CHANGES_REQUESTED** [review]

**Comment by tk0miya** [comment]
LGTM with nits

File: utils/compat_regexp_generator.py (hunk 2 above)
Comment (shibukawa) [reply utils/compat_regexp_generator.py]: done

File: utils/compat_regexp_generator.py (hunk 1 above)
Comment (shibukawa) [reply utils/compat_regexp_generator.py]: done

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
LGTM!

File: CHANGES [hunk 3]
```diff
@@ -10,6 +10,7 @@ Bugs fixed
 * #3068: Allow the '=' character in the -D option of sphinx-build.py
 * #3074: ``add_source_parser()`` crashes in debug mode
 * #3135: ``sphinx.ext.autodoc`` crashes with plain Callable
+* #3150: Fix query word splitter in JavaScript. It behaves as same as Python's regular expressio.
```
Comment (TimKam) [inline CHANGES]: "It behaves the same as Python's regular expression."

File: utils/jssplitter_generator.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+console.log('   ... ok\\n')
+
+console.log("test splitting Chinese characters")
+assert.deepEqual(['Hello', 'from', '中国', '上海'], splitQuery('Hello from 中国 上海'));
+console.log('   ... ok\\n')
+
+console.log("test splitting Emoji(surrogate pair) characters. It should keep emojis.")
+assert.deepEqual(['😁😁'], splitQuery('😁😁'));
+console.log('   ... ok\\n')
+
```
Comment (TimKam) [inline utils/jssplitter_generator.py:107]: It'd be great to add a test with umlauts here, like:

```
console.log("test splitting umlauts. It should keep umlauts.")
assert.deepEqual(['Löschen', 'Prüfung', 'Abändern', 'ærlig', 'spørsmål'], splitQuery('Löschen Prüfung Abändern ærlig spørsmål'));
console.log('   ... ok\\n')
```

**Comment by shimizukawa** [comment]
LGTM!

**Comment by tk0miya** [comment]
Okay, I will merge this tomorrow.
@shibukawa could you update this PR until tomorrow?
If impossible, I will update CHANGES entry and please send us another PR to update test for umlauts.

File: utils/jssplitter_generator.py (hunk 4 above)
Comment (shibukawa) [reply utils/jssplitter_generator.py:107]: I added. Thank you for your feedback.

**Comment by tk0miya** [comment]
Merged. Thank you always!

**Pull request merged**